rust-driver/src/types.rs

```diff
@@ -0,0 +1,451 @@
+use std::net::Ipv4Addr;
+
+use bitflags::bitflags;
+use derive_builder::Builder;
+use eui48::MacAddress;
+use num_enum::TryFromPrimitive;
+use thiserror::Error;
+
+
+/// Protection Domain
+#[derive(Debug, Clone, Copy, Default)]
+pub struct Pd {
+    handle: u32,
+}
+
+/// Type for `Imm`
+#[derive(Debug, Clone, Copy, Hash, Default)]
+pub struct Imm(u32);
+impl Imm {
+    /// Create a new `Imm` with the given value.
+    #[must_use]
+    pub fn new(imm: u32) -> Self {
+        Self(imm)
+    }
+
+    /// Get the value of `Imm`.
+    #[must_use]
+    pub fn get(&self) -> u32 {
+        self.0
+    }
+
+    /// Convert the value of `Imm` to big endian.
+    #[must_use]
+    pub fn into_be(self) -> u32 {
+        self.0.to_be()
+    }
+
+    /// Convert the value of big endian to `Imm`.
+    #[must_use]
+    pub fn from_be(val: u32) -> Self {
+        Self::new(val.to_le())
+    }
+}
+
+impl From<u32> for Imm {
+    fn from(imm: u32) -> Self {
+        Self::new(imm)
+    }
+}
+
+/// Message Sequence Number
+#[derive(Debug, Clone, Copy, Hash, PartialEq, Eq, Ord, PartialOrd)]
+pub struct Msn(u16);
+impl Msn {
+    /// Create a new `Msn` with the given value.
+    #[must_use]
+    pub fn new(msn: u16) -> Self {
+        Self(msn)
+    }
+
+    /// Get the value of `Msn`.
+    #[must_use]
+    pub fn get(&self) -> u16 {
+        self.0
+    }
+
+    /// Convert the value of `Msn` to big endian.
+    #[must_use]
+    pub fn into_be(self) -> u16 {
+        self.0.to_be()
+    }
+}
+
+impl From<u16> for Msn {
+    fn from(msn: u16) -> Self {
+        Self::new(msn)
+    }
+}
+
+impl Default for Msn {
+    fn default() -> Self {
+        Self::new(0)
+    }
+}
+
+/// `RKey` and `LKey`
+#[derive(Debug, Clone, Copy, Hash, PartialEq, Eq, Default)]
+pub struct Key(u32);
```

Comment:
I added a method to create the key from two parts and reserved the origin but renamed it with the suffix "_unchecked" because I can not check the key without the context.